One Elasticsearch index to rule them all

KatrinIhler opened the pull request:
(Is the joke getting old yet?)

With this PR, there will be only one Elasticsearch index supporting both the Admin UI and the External API. The old index implementations were removed, the new one replaced the AbstractSearchIndex class that already contained most of the relevant functionality. The index-rebuild module was merged with the elasticsearch-index module for simplification.

The endpoints to clear and rebuild the index were moved to /index. Both the Admin UI and External API endpoints related to that were removed. This might slightly violate the External API versioning, but there was no good way to keep this backwards-compatible, and you shouldn't trigger an index rebuild from an external application anyway.

The new index has a new identifier that Opencast uses to find it. When updating to OC 11, there will be three different ways to get this to work (can you tell that I couldn't decide which one was better?)

1. Just start an index rebuild. A simple way to go for smaller systems.
2. If an index rebuild will not be required for other reasons, you can use an included migration script to rename your admin UI index. There might be some things to consider about the cloning process used if you have a complex Elasticsearch setup, but for most people this should be straightforward.
3. Both too much work? If you don't care about consistency, you can use the new config option to just keep using your old admin UI index without renaming it.

Additionally, there's a second script to remove the old indices if you want to. I hope I made it sufficiently clear that that should be used with caution. 

I hope I added/updated all the relevant docs. I cleared the release notes for OC11, I think that was the right thing to do?

While I was at it, I did some very minor cleanup where unnecessary complex method parameters and unused classes were concerned. I also killed the EventIndex interface.

I added some configuration and split up the index name into name and identifier (one used for logging, one to find the actual index), because I felt they had different use cases. In general, I struggled with some indecisiveness where naming was concerned, so I hope the commits aren't too confusing. If you struggle understanding something, just ask.

PS: Forgot to add the best thing - the index rebuild should now be twice as fast ;)

KatrinIhler committed:
New Elasticsearch index implementation

This new index implementation replaces the admin UI and External API
indices. It was created in place of AbstractSearchIndex and the module
was renamed accordingly. The old index modules were removed.

Some additional minor changes:

- Some methods were moved from ApiIndex up to AbstractElasticsearchIndex
- getTermsOfField() parameters were simplified
- The EventIndex interface and implementation was removed, instead the
code lives now directly in the list providers as was already the case
for some

KatrinIhler committed:
Merge index-rebuild module with api-index module

KatrinIhler committed:
Reorganize api-index pom

KatrinIhler committed:
Move index endpoint from admin-ui to api-index module

KatrinIhler committed:
Annotate ApiIndexEndpoint

KatrinIhler committed:
Remove unused method parameter in AbstractESIndex

KatrinIhler committed:
Take hyphen out of index name

KatrinIhler committed:
Add migration scripts

KatrinIhler committed:
Update upgrade guide

KatrinIhler committed:
Add release note about new Elaticsearch index

KatrinIhler committed:
Add basic authentication to scripts

KatrinIhler committed:
Remove index endpoints of External API

...since any changes to the new index would also affect the Admin UI.
Instead, there's now a new endpoint at /index.

KatrinIhler committed:
Change index name to simply 'index'

KatrinIhler committed:
Update documentation

KatrinIhler committed:
Rename ApiIndex to ElasticsearchIndex

KatrinIhler committed:
Improve migration scripts

KatrinIhler committed:
Make index name configurable

KatrinIhler committed:
Change index name to 'opencast'

KatrinIhler committed:
Add option 3 to upgrade guide

KatrinIhler committed:
Remove trailing spaces

KatrinIhler committed:
Differentiate between index identifier and name

KatrinIhler committed:
Delete unused classes

KatrinIhler committed:
Minor fixes

KatrinIhler committed:
Correct docs

KatrinIhler committed:
Fix config parsing bug

KatrinIhler committed:
Harmonize logging

KatrinIhler committed:
Fix trailing spaces AGAIN

lkiesow commented on docs/guides/admin/docs/releasenotes/filter-admin-ui-users.txt:
```diff
@@ -1 +0,0 @@
-- [#2019](https://github.com/opencast/opencast/pull/2019): The full list of Opencast users can now be reduced in the admin UI's filtering tool using a regular expression
```
Why does this delete any release notes?

KatrinIhler replied:
I was already wondering whether that was the right way to go - do we want to keep these for older releases? If I remember correctly, the extra files are just for the release managers so they can easily create the release notes, so having them in there for multiple releases without distinction might be very confusing? On the other hand, these are extra files so merging forward is easier, if I remember correctly, so... Idk, I'm confused. These wouldn't be new features in OC11, so do we want to keep them in here or not? I'm fine with either.

lkiesow commented on docs/guides/admin/docs/upgrade.md:
```diff
@@ -1,70 +1,75 @@
+Example (using cURL):
 
-Finally, check you have the correct version using:
+    curl -i --digest -u <digest_user>:<digest_password> -H "X-Requested-Auth: Digest" -s -X POST \
```
```suggestion
    curl -i -u <admin_user>:<password> -s -X POST \
```

lkiesow commented on docs/upgrade/10_to_11/delete-old-indices.sh:
```diff
@@ -0,0 +1,63 @@
+
+		# Make sure index is open
+		echo -e "Making sure index is open:\n"
+		curl -u $USER:$PASSWORD -X POST  "$SERVER/$DELETE_INDEX/_open" -w "\n\n" -i -s
```
```suggestion
		curl -u "$USER:$PASSWORD" -X POST  "$SERVER/$DELETE_INDEX/_open" -w "\n\n" -i -s
```

lkiesow commented on docs/upgrade/10_to_11/delete-old-indices.sh:
```diff
@@ -0,0 +1,63 @@
+
+		# Delete index
+		echo -e "Deleting index:\n"
+		RESPONSE=$(curl -u $USER:$PASSWORD -X DELETE  "$SERVER/$DELETE_INDEX" -i -s)
```
```suggestion
		RESPONSE=$(curl -u "$USER:$PASSWORD" -X DELETE  "$SERVER/$DELETE_INDEX" -i -s)
```

lkiesow commented on docs/upgrade/10_to_11/delete-old-indices.sh:
```diff
@@ -0,0 +1,63 @@
+THEME="theme"
+GROUP="group"
+
+set -e
```
```suggestion
set -eu
```

lkiesow commented on docs/upgrade/10_to_11/delete-old-indices.sh:
```diff
@@ -0,0 +1,63 @@
+		echo -e "${RESPONSE}\n"
+
+	        # Check for errors
+		if [[ $RESPONSE != *"200"* ]]; then
```
Maybe use `curl -f …`?

KatrinIhler replied:
`-f` swallows the response body, and I thought that information might be helpful to some people. There is `--fail-with-body` now, but it's very new, my cURL for example can't do that yet...

lkiesow commented on docs/upgrade/10_to_11/delete-old-indices.sh:
```diff
@@ -0,0 +1,63 @@
+# Delete old indices that will no longer be used in Opencast >= 11
+# Make sure you use the migration script first if you don't want to do an index rebuild!
+echo -e "Indices before:\n"
+curl -u $USER:$PASSWORD -X GET "$SERVER/_cat/indices?v" -w "\n"
```
```suggestion
curl -u "$USER:$PASSWORD" -X GET "$SERVER/_cat/indices?v" -w "\n"
```

lkiesow commented on docs/upgrade/10_to_11/delete-old-indices.sh:
```diff
@@ -0,0 +1,63 @@
+echo -e "Cleanup done.\n"
+
+echo -e "Indices after:\n"
+curl -u $USER:$PASSWORD -X GET "$SERVER/_cat/indices?v" -w "\n"
```
```suggestion
curl -u "$USER:$PASSWORD" -X GET "$SERVER/_cat/indices?v" -w "\n"
```

lkiesow commented on docs/upgrade/10_to_11/migrate-indices.sh:
```diff
@@ -0,0 +1,80 @@
+SERIES="series"
+THEME="theme"
+
+set -e
```
```suggestion
set -eu
```

Plus all the `$USER:$PASSWORD` should be in quotes

lkiesow reviewed (dismissed):
Faster index rebuilds and less duplication :tada:

KatrinIhler commented:
@lkiesow Thanks for the review, I'm gonna do the necessary changes sometime next week! :)

gregorydlogan commented:
> I hope I added/updated all the relevant docs. I cleared the release notes for OC11, I think that was the right thing to do?

This seems reasonable to me.  In theory this is the job of the RM, but we don't have one formally appointed yet, and I'm not sure it matters anyway :)

KatrinIhler commented:
Fix for the database test failure is waiting for review at #2877. Will rebase once that's merged. :)

KatrinIhler committed:
Use basic auth in request example in docs

KatrinIhler committed:
Fix migration scripts

KatrinIhler commented:
@lkiesow I think I addressed all of your review findings now. :)

gregorydlogan commented on docs/guides/admin/docs/upgrade.md:
```diff
@@ -6,10 +6,62 @@ In case you need information about how to upgrade older versions of Opencast,
+In Opencast 11, instead of having two separate Elasticsearch indices for the Admin UI and the External API that share
+most of their content, there will be only one index supporting both. Since this index will have a new name, this index
+will either have to be rebuilt from scratch, or you can migrate your Admin UI index, since its structure is completely
+identical to the new index. However, if you don't care about accuracy, you could just keep using the Admin UI index.
```
So, can I just do nothing and ignore these instructions completely?  My read of the previous sentence is that I need to either rebuild, or move/migrate my existing index.  The last sentence implies that I can do nothing, and suffer from reduced functionality in the external API (ie, its index is missing), but otherwise stuff will work as it previously did.

KatrinIhler replied:
You can't do _nothing_, exactly. If you want to keep using the Admin UI index without migrating it, you would have to change the index name in the config accordingly, as is described in Option 3. Then, the External API will also use the Admin UI index (which is why it's the last option, since this is a bit unclean).

If you don't do anything, Opencast would create an empty index under the new name for both, so then you basically have to do a rebuild.

gregorydlogan commented on docs/guides/admin/docs/upgrade.md:
```diff
@@ -6,10 +6,62 @@ In case you need information about how to upgrade older versions of Opencast,
-2. TBA
+2. Replace Opencast with the new version
+3. Review the configuration changes and adjust your configuration accordingly
+4. [Migrate (or rebuild) Elasticsearch index](#migrate-elasticsearch)
```
This step needs to be done *after* starting Opencast if you're rebuilding, right?

Perhaps
```
please refer to [older release notes](...)

Bold font: Note: Only one of steps 4, or 6 are required.

1. Stop your current Opencast install
...
4. Optionally [Migrate Elasticsearch index]... -> linking directly to the relevant part
5. Start Opencast
6. Optionally [Rebuild Elasticsearch index]... -> linking directly to the relevant part
```

KatrinIhler replied:
Only the rebuild needs to happen during runtime, however, Option 1 clearly states this.
I put this before the restart since people should read the whole thing and decide what they want to do before starting Opencast. So I'd rather not link to the options directly. However, I'm gonna make the instructions a little more clear.

gregorydlogan commented on docs/guides/admin/docs/upgrade.md:
```diff
@@ -6,10 +6,62 @@ In case you need information about how to upgrade older versions of Opencast,
+
+### Option 1: Rebuild
+
+Start your new Opencast and make an HTTP POST request to `/index/rebuild`.
```
Changes to step list means we can remove the Starting of Opencast I think.

KatrinIhler replied:
Added an 'if you haven't already done so' to make things clearer.

gregorydlogan commented on docs/guides/admin/docs/upgrade.md:
```diff
@@ -6,10 +6,62 @@ In case you need information about how to upgrade older versions of Opencast,
+This will clone the Admin UI index to an index with the new name. If your storage supports hardlinks, the cloning
+process should be pretty quick.
+
+If you want to clean up the old indices afterwards, you can use `delete-indices.sh` for this.
```
IMO this should be moved down to the important notices section below.  Don't tell people there's a cleanup script before telling them why they shouldn't run it without testing things first :)

KatrinIhler replied:
Done.

KatrinIhler replied:
Changed the order so the important notices come first. :)

gregorydlogan commented on docs/guides/admin/docs/upgrade.md:
```diff
@@ -6,10 +6,62 @@ In case you need information about how to upgrade older versions of Opencast,
+
+- Please take a quick look at the scripts before using them and adjust them if needed.
+- Do not start your new Opencast before migrating the index! This will create a new empty index which will cause the
+  cloning process to fail. To fix this, delete all subindices of the new index (event, series, theme, version) before
```
I know where the subindices live, but I would probably give a hint on how to do this.

KatrinIhler replied:
Referenced the deletion script for the relevant commands. Made really sure people don't just run it...

gregorydlogan commented on docs/guides/admin/docs/upgrade.md:
```diff
@@ -6,10 +6,62 @@ In case you need information about how to upgrade older versions of Opencast,
+- Do not start your new Opencast before migrating the index! This will create a new empty index which will cause the
+  cloning process to fail. To fix this, delete all subindices of the new index (event, series, theme, version) before
+  attempting migration again.
+- The old indices cannot be used during migration, so at least the Admin node shouldn't be running.
```
This, taken with the `Do not start` line above should, imo, be combined into something like:

`- The migration process is an offline process that cannot be safely accomplished while Opencast is running.  Ensure Opencast is stopped on the admin node prior to starting this migration.`

KatrinIhler replied:
Attempted to do this, but damn, is it hard to make this concise, but safe.

gregorydlogan commented on docs/guides/admin/docs/upgrade.md:
```diff
@@ -6,10 +6,62 @@ In case you need information about how to upgrade older versions of Opencast,
+you could also just set `index.identifier` in `org.opencastproject.elasticsearch.index.ElasticsearchIndex` to "adminui"
+to keep using the old admin ui index. Do this before starting Opencast to avoid creating a new index.
+
+The external API index can then be removed. (Please be aware that the External API index cannot be used any longer since
```
How does one do this?  Same script as above?

KatrinIhler replied:
Added explanation.

gregorydlogan reviewed (changes requested):
LGTM, minor doc nitpicks.  Also ran into this:
```
2021-09-29T15:16:19,499 | INFO  | (IndexEndpoint:112) - Clear the index
2021-09-29T15:16:29,976 | INFO  | (IndexEndpoint:158) - Starting to repopulate the index
2021-09-29T15:16:30,235 | ERROR | (AbstractEventEndpoint:2435) - The admin UI Search Index was not able to get the events list:
org.opencastproject.elasticsearch.api.SearchIndexException: Error querying event index
	at org.opencastproject.elasticsearch.index.ElasticsearchIndex.getByQuery(ElasticsearchIndex.java:584) ~[!/:?]
	at org.opencastproject.adminui.endpoint.AbstractEventEndpoint.getEvents(AbstractEventEndpoint.java:2433) [!/:?]
	at jdk.internal.reflect.GeneratedMethodAccessor84.invoke(Unknown Source) ~[?:?]
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:566) ~[?:?]
...
Caused by: org.opencastproject.elasticsearch.api.SearchIndexException: [opencast_event] ElasticsearchStatusException[Elasticsearch exception [type=index_not_found_exception, reason=no such index [opencast_event]]]
	at org.opencastproject.elasticsearch.impl.AbstractElasticsearchIndex.executeQuery(AbstractElasticsearchIndex.java:536) ~[!/:?]
	at org.opencastproject.elasticsearch.index.ElasticsearchIndex.getByQuery(ElasticsearchIndex.java:576) ~[!/:?]
	... 115 more
Caused by: org.elasticsearch.ElasticsearchStatusException: Elasticsearch exception [type=index_not_found_exception, reason=no such index [opencast_event]]
	at org.elasticsearch.rest.BytesRestResponse.errorFromXContent(BytesRestResponse.java:187) ~[!/:?]
	at org.elasticsearch.client.RestHighLevelClient.parseEntity(RestHighLevelClient.java:1911) ~[!/:?]
	at org.elasticsearch.client.RestHighLevelClient.parseResponseException(RestHighLevelClient.java:1888) ~[!/:?]
	at org.elasticsearch.client.RestHighLevelClient.internalPerformRequest(RestHighLevelClient.java:1645) ~[!/:?]
	at org.elasticsearch.client.RestHighLevelClient.performRequest(RestHighLevelClient.java:1602) ~[!/:?]
	at org.elasticsearch.client.RestHighLevelClient.performRequestAndParseEntity(RestHighLevelClient.java:1572) ~[!/:?]
	at org.elasticsearch.client.RestHighLevelClient.search(RestHighLevelClient.java:1088) ~[!/:?]
	at org.opencastproject.elasticsearch.impl.AbstractElasticsearchIndex.executeQuery(AbstractElasticsearchIndex.java:534) ~[!/:?]
	at org.opencastproject.elasticsearch.index.ElasticsearchIndex.getByQuery(ElasticsearchIndex.java:576) ~[!/:?]
	... 115 more
	Suppressed: org.elasticsearch.client.ResponseException: method [POST], host [http://localhost:9200], URI [/opencast_event/_search?typed_keys=true&max_concurrent_shard_requests=5&ignore_unavailable=false&expand_wildcards=open&allow_no_indices=true&preference=_local&ignore_throttled=true&search_type=query_then_fetch&batched_reduce_size=512&ccs_minimize_roundtrips=true], status line [HTTP/1.1 404 Not Found]
{"error":{"root_cause":[{"type":"index_not_found_exception","reason":"no such index [opencast_event]","resource.type":"index_or_alias","resource.id":"opencast_event","index_uuid":"_na_","index":"opencast_event"}],"type":"index_not_found_exception","reason":"no such index [opencast_event]","resource.type":"index_or_alias","resource.id":"opencast_event","index_uuid":"_na_","index":"opencast_event"},"status":404}
		at org.elasticsearch.client.RestClient.convertResponse(RestClient.java:326) ~[!/:?]
		at org.elasticsearch.client.RestClient.performRequest(RestClient.java:296) ~[!/:?]
		at org.elasticsearch.client.RestClient.performRequest(RestClient.java:270) ~[!/:?]
		at org.elasticsearch.client.RestHighLevelClient.internalPerformRequest(RestHighLevelClient.java:1632) ~[!/:?]
		at org.elasticsearch.client.RestHighLevelClient.performRequest(RestHighLevelClient.java:1602) ~[!/:?]
		at org.elasticsearch.client.RestHighLevelClient.performRequestAndParseEntity(RestHighLevelClient.java:1572) ~[!/:?]
		at org.elasticsearch.client.RestHighLevelClient.search(RestHighLevelClient.java:1088) ~[!/:?]
		at org.opencastproject.elasticsearch.impl.AbstractElasticsearchIndex.executeQuery(AbstractElasticsearchIndex.java:534) ~[!/:?]
		at org.opencastproject.elasticsearch.index.ElasticsearchIndex.getByQuery(ElasticsearchIndex.java:576) ~[!/:?]
		at org.opencastproject.adminui.endpoint.AbstractEventEndpoint.getEvents(AbstractEventEndpoint.java:2433) [!/:?]
		at jdk.internal.reflect.GeneratedMethodAccessor84.invoke(Unknown Source) ~[?:?]
		at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
		at java.lang.reflect.Method.invoke(Method.java:566) ~[?:?]
```

To trigger this I started a fresh OC 11 build, processed a single recording, cleared the index, then rebuilt via the rest docs.

Edit: Also, the reindex *still appears to work* since the event reappeared in the admin page after the reindex ¯\\_(ツ)_/¯

KatrinIhler committed:
Small fix

KatrinIhler committed:
Make upgrade guide clearer

KatrinIhler commented:
@gregorydlogan I think I addressed all your review findings now. Writing upgrade docs for something like this is hard... but as long as people actually read the damn thing, they should be fine.
I could not reproduce that Exception of yours even though I did the exact same things in the same order. This almost looks like an Elasticsearch glitch? I mean, clearly _some_ event index was available before. This could only happen if the Elasticsearch was restarted and doesn't keep its state (my docker setup does this) or if you would change the index identifier in the middle of things, which can't happen without a module restart (or maybe even OC restart). Very odd... We're running this code at SWITCH since the summer and haven't seen something like this yet.

mliradelc commented:
All issues are addressed, and is proven in production since summer.

Thanks for the awesome job 😄

Merged!
